test/functional/testlib.bash

```diff
@@ -0,0 +1,502 @@
+export FUNC_DIR
+export SWUPD_DIR="$FUNC_DIR/../.."
+export SWUPD="$SWUPD_DIR/swupd"
+
+
+generate_random_content() { 
+
+	local bottom_range=5
+	local top_range=100
+	local range=$((top_range - bottom_range + 1))
+	local number_of_lines=$((RANDOM%range))  
+	< /dev/urandom tr -dc 'a-zA-Z0-9-_!@#$%^&*()_+{}|:<>?=' | fold -w 100 | head -n $number_of_lines
+
+}
+
+generate_random_name() { 
+
+	local prefix=${1:-test-}
+	local uuid
+	
+	# generate random 8 character alphanumeric string (lowercase only)
+	uuid=$(< /dev/urandom tr -dc 'a-f0-9' | fold -w 8 | head -n 1)
+	echo "$prefix$uuid"
+
+}
+
+validate_path() { 
+
+	local path=$1
+	if [ -z "$path" ] || [ ! -d "$path" ]; then
+		echo "Please provide a valid path"
+		return 1
+	fi
+
+}
+
+validate_item() { 
+
+	local vfile=$1
+	if [ -z "$vfile" ] || [ ! -e "$vfile" ]; then
+		echo "Please provide a valid file"
+		return 1
+	fi
+
+}
+
+validate_param() {
+
+	local param=$1
+	if [ -z "$param" ]; then
+		echo "Mandatory parameter missing"
+		return 1
+	fi
+
+}
+
+# Exports environment variables that are dependent on the test environment
+# Parameters:
+# - ENV_NAME: the name of the test environment
+set_env_variables() {
+
+	local env_name=$1
+	validate_path "$env_name"
+
+	export SWUPD_OPTS="-S$BATS_TEST_DIRNAME/$env_name/state -p $BATS_TEST_DIRNAME/$env_name/target-dir -F staging -u file://$BATS_TEST_DIRNAME/$env_name/web-dir -C $FUNC_DIR/Swupd_Root.pem -I"
+
+}
+
+# Creates a directory with a hashed name in the specified path, if a directory
+# already exists it returns the name
+# Parameters:
+# - PATH: the path where the directory will be created 
+create_dir() { 
+	
+	local path=$1
+	local hashed_name
+	local directory
+	validate_path "$path"
+	
+	# most directories have the same hash, so we only need one directory
+	# in the files directory, if there is already one just return the path/name
+	directory=$(find "$path"/* -type d 2> /dev/null)
+	if [ ! "$directory" ]; then
+		sudo mkdir "$path"/testdir
+		hashed_name=$(sudo "$SWUPD" hashdump "$path"/testdir 2> /dev/null)
+		sudo mv "$path"/testdir "$path"/"$hashed_name"
+		# since tar is all we use, create a tar for the new dir
+		create_tar "$path"/"$hashed_name"
+		directory="$path"/"$hashed_name"
+	fi
+	echo "$directory"
+
+}
+
+# Generates a file with a hashed name in the specified path
+# Parameters:
+# - PATH: the path where the file will be created    
+create_file() {
+ 
+	local path=$1
+	local hashed_name
+	validate_path "$path"
+
+	generate_random_content | sudo tee "$path/testfile" > /dev/null
+	hashed_name=$(sudo "$SWUPD" hashdump "$path"/testfile 2> /dev/null)
+	sudo mv "$path"/testfile "$path"/"$hashed_name"
+	# since tar is all we use, create a tar for the new file
+	create_tar "$path"/"$hashed_name"
+	echo "$path/$hashed_name"
+
+}
+
+# Creates a symbolic link with a hashed name to the specified file in the specified path.
+# If no existing file is specified to point to, a new file will be created and pointed to
+# by the link.
```

Comment:
Seems a little restrictive, this means it can not create dangling links.

Reply:
@icarus-sparry  what is a dangling link? 

This function is basically used by the create_bundle function to create a link (and its associated file) if a user chooses to add a link to its test bundle. I don't know if a dangling link is really necessary for a test bundle, I haven't even seen any test that currently uses a bundle that contains a link, I added the capability of adding links to make it more flexible